### ISSUE TITLE
[0.17] Need legit way to set getCanonicalName for SystemJSNodeLoader

### ISSUE THREAD

**Issue by Rush** [jspm/jspm-cli#2202]
I am trying to pre-compile JS for Babel on server-side using a custom middleware. I can do it manually with node-installed babel transpiling the codes to CJS. However, optimally I'd like to assure I'm doing it exactly the same as JSPM would do. 

```js
// running in node.js
import jspm from 'jspm';

jspm.import('plugin-babel').then(babel => {
  console.log('GOT BABEL', babel);
  
  babel.translate({
    name: 'foo', metadata: {
      format: 'esm'
    }, address: 'aaaa', source: 'export let foo = "foo"; export async function foo2() {} '
  }, {
    outputESM: false
  }).then(translated => {
    console.log('TRANSLATED', translated);
  });
})
```

yields:
```js
System.register(["file:///home/rush/code/frontend/javascripts/jspm_packages/npm/systemjs-plugin-babel@0.0.17/regenerator-runtime.js", "file:///home/rush/code/frontend/javascripts/jspm_packages/npm/systemjs-plugin-babel@0.0.17/babel-helpers/asyncToGenerator.js"], function (_export, _context) {
  "use strict";

  var _regeneratorRuntime, _asyncToGenerator, foo, foo2;

  return {
    setters: [function (_fileHomeRushFrontendJavascriptsJspm_packagesNpmSystemjsPluginBabel0017RegeneratorRuntimeJs) {
      _regeneratorRuntime = _fileHomeRushFrontendJavascriptsJspm_packagesNpmSystemjsPluginBabel0017RegeneratorRuntimeJs.default;
    }, function (_fileHomeRushFrontendJavascriptsJspm_packagesNpmSystemjsPluginBabel0017BabelHelpersAsyncToGeneratorJs) {
      _asyncToGenerator = _fileHomeRushFrontendJavascriptsJspm_packagesNpmSystemjsPluginBabel0017BabelHelpersAsyncToGeneratorJs.default;
    }],
    execute: function () {
      _export("foo", foo = "foo");

      _export("foo", foo);

      _export("foo2", foo2 = function () {
        var _ref = _asyncToGenerator(_regeneratorRuntime.mark(function _callee() {
          return _regeneratorRuntime.wrap(function _callee$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                case "end":
                  return _context2.stop();
              }
            }
          }, _callee, this);
        }));

        return function foo2() {
          return _ref.apply(this, arguments);
        };
      }());

      _export("foo2", foo2);
    }
  };
});
```

Obviously it won't work due to `file://` paths. Any tips?

**Comment by Rush** [comment]
I am gonna answer myself:
```js
let loader = new jspm.Loader({});
let builder = new jspm.Builder({});

loader.getCanonicalName = builder.getCanonicalName;
  
loader.import('plugin-babel').then(async babel => {
  let metadata = {};
  // loader.packages = {};
  // loader.map = {};    
  
  babel.translate({
    name: 'foo', metadata: {
      format: 'esm'
    }, address: 'aaaa', source: `import 'angular'; export let foo = "foo"; export async function foo2() {} `
    
  }, {
    outputESM: false
  }).then(translated => {
    console.log('TRANSLATED', translated);
  });
});
```

Not very clean but works well!

I can also get custom `babelOptions` with:
```js
let name = await loader.locate({name: pathname, metadata}); 
```

@guybedford Can this part `loader.getCanonicalName = builder.getCanonicalName;` be mainstreamed?

This is the part I am hacking around: https://github.com/systemjs/plugin-babel/blob/master/plugin-babel.js#L23

**Comment by Rush** [comment]
And yet again fixing my own answer:
```js
let loader = new jspm.Loader({});

// would be nice to have a proper solution to this workaround
loader.getCanonicalName = (new jspm.Builder).getCanonicalName;

let metadata = {};
// loader.locate can be run here to get proper metadata

loader.translate({
  name: 'foo', metadata: {
    format: 'esm'
  }, address: 'aaaa', source: `import 'angular'; export let foo = "foo"; export async function foo2() {} `
  
}, {
  outputESM: false
}).then(translated => {
  console.log('TRANSLATED', translated);
});
```

Production module using the above approach is here: https://github.com/virtkick/node-serve-jspm

**Comment by Rush** [comment]
This is broken due to changes in 0.20 https://github.com/systemjs/systemjs/issues/1554

**Comment by Rush** [comment]
Not needed due to closed systemjs/systemjs#1554
